exercises/practice/nucleotide-count/runner.mips

```diff
@@ -0,0 +1,160 @@
+# - (if it uses s0-7 then it is responsible for pushing existing values to the stack then popping them back off before returning)
+# - write counts for A,C,G,T to the word array with address given in a1
+# - write -1 values to the array if the input is invalid
+
+.data
+
+# number of test cases
+n: .word 5
+# input values and expected output values (all null terminated)
+ins:  .asciiz
+        "",
+        "G",
+        "GGGGGGG",
+        "AGCTTTTCATTCTGACTGCAACGGGCAATATGTCTCTGTGTGGATTAAAAAAAGAGTGTCTGATAGCAGC",
+        "AGXXACT"
+
+outs: .word
+        0, 0, 0, 0,
+        0, 0, 1, 0,
+        0, 0, 7, 0,
+        20, 12, 17, 21,
+        -1, -1, -1, -1
+
+failmsg: .asciiz "failed for test input: "
+expectedmsg: .asciiz ". expected "
+tobemsg: .asciiz " to be "
+okmsg: .asciiz "all tests passed"
+
+
+.text
+
+runner:
+        lw      $s0, n
+        la      $s1, ins
+        la      $s2, outs
+
+        li      $v0, 9                  # code for allocating heap memory
+        li      $a0, 16                 # specify 16 bytes - size of array
+        syscall
+        move    $s5, $v0                # location of allocated memory is where callee writes result
+
+run_test:
+        jal     clear_output            # zero out output location
+        move    $a0, $s1                # load input value into a0
+        move    $a1, $s5
+        jal     nucleotide_counts       # call subroutine under test
+
+        lw      $t0, 0($s2)
+        lw      $t1, 0($s5)
+        bne     $t0, $t1, exit_fail     # Compare A counts
+
+        lw      $t0, 4($s2)
+        lw      $t1, 4($s5)
+        bne     $t0, $t1, exit_fail     # Compare C counts
+
+        lw      $t0, 8($s2)
+        lw      $t1, 8($s5)
+        bne     $t0, $t1, exit_fail     # Compare G counts
+
+        lw      $t0, 12($s2)
+        lw      $t1, 12($s5)
+        bne     $t0, $t1, exit_fail     # Compare T counts
+
+input_scan:
+        lb      $s3, 0($s1)
+        addi    $s1, $s1, 1
+        bne     $s3, $zero, input_scan
+
+done_scan:
+        addi    $s2, $s2, 16
+        subi    $s0, $s0, 1             # decrement num of tests left to run
+        bgt     $s0, $zero, run_test    # if more than zero tests to run, jump to run_test
+
+exit_ok:
+        la      $a0, okmsg              # put address of okmsg into a0
+        li      $v0, 4                  # 4 is print string
+        syscall
+
+        li      $v0, 10                 # 10 is exit with zero status (clean exit)
+        syscall
+
+exit_fail:
+        la      $a0, failmsg            # put address of failmsg into a0
+        li      $v0, 4                  # 4 is print string
+        syscall
+
+        move    $a0, $s1                # print input that failed on
+        li      $v0, 4
+        syscall
+
+        la      $a0, expectedmsg
+        li      $v0, 4
+        syscall
+
+        lw      $a0, 0($s5)             # print actual A that failed on
```

Comment:
Lines 112..126 and 132..146 could be extracted into a subroutine, with $s5 or $s2 passed as an argument, perhaps using $a2.

We could print `"[1, 2, 3, 4]"` instead of the current `"1234"`.

Reply:
I feel like 1234 is fine for now, but feel free to change it if you like

Reply:
Changed. The all-your-base runner will also have a subroutine, to print lists like `"[6, 10, 45]"`